core/src/main/java/com/linecorp/armeria/internal/common/websocket/WebSocketFrameDecoder.java

```diff
@@ -0,0 +1,370 @@
+                        protocolViolation(WebSocketCloseStatus.MESSAGE_TOO_BIG,
+                                          "Max frame length of " + maxFramePayloadLength +
+                                          " has been exceeded.");
+                        return;
+                    }
+
+                    logger.trace("Decoding a WebSocket Frame. length: {}", framePayloadLength);
+
+                    state = State.MASKING_KEY;
+                    // fallthrough
+                case MASKING_KEY:
+                    if (frameMasked) {
+                        if (in.readableBytes() < 4) {
+                            return;
+                        }
+                        mask = in.readInt();
+                    }
+                    state = State.PAYLOAD;
+                    // fallthrough
+                case PAYLOAD:
+                    if (in.readableBytes() < framePayloadLength) {
+                        return;
+                    }
+
+                    ByteBuf payloadBuffer = Unpooled.EMPTY_BUFFER;
+                    if (framePayloadLength > 0) {
+                        payloadBuffer = in.readBytes(toFrameLength(framePayloadLength));
+                    }
+
+                    // Now we have all the data, the next checkpoint must be the next
+                    // frame
+                    state = State.READING_FIRST;
+
+                    // Unmask data if needed
+                    if (frameMasked & framePayloadLength > 0) {
+                        unmask(payloadBuffer);
+                    }
+
+                    // Processing ping/pong/close frames because they cannot be
+                    // fragmented
+                    if (frameOpcode == WebSocketFrameType.PING.opcode()) {
+                        final WebSocketFrame decodedFrame = WebSocketFrame.ofPooledPing(payloadBuffer);
+                        out.add(decodedFrame);
+                        logger.trace("{} is decoded.", decodedFrame);
+                        continue; // to while loop
+                    }
+                    assert payloadBuffer != null;
+                    if (frameOpcode == WebSocketFrameType.PONG.opcode()) {
+                        final WebSocketFrame decodedFrame = WebSocketFrame.ofPooledPong(payloadBuffer);
+                        out.add(decodedFrame);
+                        logger.trace("{} is decoded.", decodedFrame);
+                        continue; // to while loop
+                    }
+                    if (frameOpcode == WebSocketFrameType.CLOSE.opcode()) {
+                        receivedClosingHandshake = true;
+                        validateCloseFrame(payloadBuffer);
+                        final CloseWebSocketFrame decodedFrame = WebSocketFrame.ofPooledClose(payloadBuffer);
+                        out.add(decodedFrame);
+                        logger.trace("{} is decoded.", decodedFrame);
+                        webSocketCloseHandler.receivedCloseFrame();
```

Comment:
I understand it is complicated to handle the input stream at `WebSocketCloseHandler` layer. However, `WebSocketUtil.closeWebSocketInboundStream()` looks like a hacky way to close the input stream. Is it possible to get the same effect if we just call `in.close()` when `WebSocketFrameType.CLOSE` is received? If `in.close()` is called, will input will be automatically discarded. 
https://github.com/line/armeria/blob/ecc0e89b99b486709722210ac4fcbe6c77764a56/core/src/main/java/com/linecorp/armeria/internal/common/stream/ByteBufsDecoderInput.java#L46-L49

Reply:
This was the trickiest part. 😄 
`closeWebSocketInboundStream` is also called from `WebSocketResponseSubscriber` which cannot access the `in`. If I managed to do that, we still cannot close the original `HttpRequest`. 

So I thought this was the only way to close the original `HttpRequest` for now. But probably we don't have to close the original request and just let it be aborted when the `HttpResponse` is complete.
If it's acceptable please let me know.
I might create an `HttpRequest` in `WebSocketService` and pass the elements to the request from the original `HttpRequest` to also close the request from `WebSocketResponseSubscriber`.